bdk_esplora_example/src/esplora.rs

```diff
@@ -0,0 +1,188 @@
+use bdk_core::{
+    bitcoin::{BlockHash, Script, Transaction},
+    chain_graph::ChainGraph,
+    keychain::KeychainScan,
+    sparse_chain::{InsertCheckpointErr, InsertTxErr},
+    BlockId, ConfirmationTime,
+};
+use std::collections::BTreeMap;
+
+use esplora_client::{BlockingClient, Builder};
+
+use bdk_cli::Result;
+
+#[derive(Debug, Clone)]
+pub struct Client {
+    pub parallel_requests: u8,
+    pub client: BlockingClient,
+}
+
+impl Client {
+    /// TODO
+    pub fn new(base_url: &str, parallel_requests: u8) -> Result<Self> {
+        Ok(Self {
+            parallel_requests,
+            client: Builder::new(base_url).build_blocking()?,
+        })
+    }
+
+    pub fn spk_scan(
+        &self,
+        spks: impl Iterator<Item = Script>,
+        existing_chain: BTreeMap<u32, BlockHash>,
+    ) -> Result<ChainGraph<ConfirmationTime>> {
+        let mut dummy_keychains = BTreeMap::new();
+        dummy_keychains.insert((), spks.enumerate().map(|(i, spk)| (i as u32, spk)));
+
+        let wallet_scan = self.wallet_scan(dummy_keychains, None, existing_chain)?;
+
+        Ok(wallet_scan.update)
+    }
+
+    /// Create a new checkpoint with transactions spending from or to the scriptpubkeys in
+    /// `scripts`.
+    pub fn wallet_scan<K: Ord + Clone, I>(
+        &self,
+        keychains: BTreeMap<K, I>,
+        stop_gap: Option<usize>,
```

Comment:
This is out of the scope of this PR, but I am rather curious about the definition of `stop_gap` here.

In my mind, spks should be dynamically added and stop gap is used to determine when to stop adding spks in which we should scan for.

The implementation here is first creating a large batch of spks, and stop gap is used to determine when to stop scanning.

In my opinion, scanning with dynamically-added-spks would be more useful.

Reply:
Thats a nice idea.. But one caveat could be, IIUC that we don't wanna take in `refs` of the tracker, because of some async issues occurred in bdk prior (as per @LLFourn ).. So we need to have the script source from somewhere to do dynamic feeding.. The electrum should not have to derive scripts..

Maybe some cool loop with feedback system can be designed to dynamically feed scripts into the `scan`, until it hits `stop_gap` amount of `unused` `sync_response`? Something along that line..

But for regular wallet usages with even 1k `used` scripts per keychain, passing on everything in the tracker for sync should not so terrible, in exchange for simpler code.. 

Also note the script iterator size is maxed at the total `derived index` of the tracker.. So a design goal for the tracker should be to be efficient in storing derived indexes, and not have large pool of unused derived scripts..

Reply:
> This is out of the scope of this PR, but I am rather curious about the definition of `stop_gap` here.

Same definition as is always used.

> The implementation here is first creating a large batch of spks, and stop gap is used to determine when to stop scanning.

This is not what is happening. The API is an `Iterator`.

> Maybe some cool loop with feedback system can be designed to dynamically feed scripts into the `scan`, until it hits `stop_gap` amount of `unused` `sync_response`? Something along that line..

This cool feedback system is called an `Iterator` which is what we already doing :).

Reply:
@LLFourn good point, sorry I misread the code.